neotest-kotlin/app/src/main/kotlin/io/github/codymikol/neotestkotlin/framework/TestNode.kt

```diff
@@ -0,0 +1,182 @@
+                    val container = queue.removeFirst()
+
+                    container.nodes.forEach { child ->
+                        yield(child)
+
+                        if (child is TestNode.Container) {
+                            queue.add(child)
+                        }
+                    }
+                }
+            }
+        }
+
+        /**
+         * Adds the [node] to the [Container] appending onto potentially nested [TestNode.Container]s
+         * specified by [parentNames]. An empty list for [parentNames] means that it's top level.
+         *
+         * @throws IllegalArgumentException [parentNames] references a non-existent node or a node that
+         * isn't a [TestNode.Container].
+         */
+        @Throws(IllegalArgumentException::class)
+        fun add(
+            node: TestNode,
+            parentNames: List<String> = emptyList(),
+        ): Boolean {
+            val child =
+                parentNames.fold(this) { currentNode, parentName ->
+                    val child =
+                        requireNotNull(currentNode.nodes.find { it.name == parentName }) {
+                            "${this.name} > ${parentNames.joinToString(separator = " > ")} does not exist"
+                        }
+
+                    require(child is TestNode.Container) {
+                        "${this.name} > ${parentNames.joinToString(separator = " > ")} is not a TestNode.Container"
+                    }
+
+                    child
+                }
+
+            require(
+                !child.nodes.any { it.name == node.name },
+            ) { "${this.name} > ${parentNames.joinToString(separator = " > ")} > ${node.name} already exists" }
+
+            return child.nodes.add(node)
+        }
+    }
+}
+
+@Serializable
+enum class Status {
+    SUCCESS,
+    FAILURE,
+    IGNORED,
+}
+
+@Serializable
+@OptIn(ExperimentalSerializationApi::class) // JsonClassDiscriminator is an experimental annotation for now
+@JsonClassDiscriminator("status")
+sealed interface TestStatus {
+    @Serializable
+    @SerialName("success")
+    object Success : TestStatus
+
+    companion object {
+        fun from(kotestResult: TestResult): TestStatus =
+            when (kotestResult) {
+                is TestResult.Success -> TestStatus.Success
+                is TestResult.Failure -> {
+                    val error = kotestResult.errorOrNull
+
+                    TestStatus.Failure(
+                        stackTrace = error?.stackTraceToString(),
+                        error =
+                            error?.let {
+                                val traceOrigin = it.stackTrace?.firstOrNull()
+
+                                Failure.Error(
+                                    message = it.message,
+                                    lineNumber = traceOrigin?.lineNumber,
+                                    filename = traceOrigin?.fileName,
+                                )
```

Comment:
This mostly works, outside of soft assertions which seem to reference a lineNumber that is higher than the file. Maybe just need to specially handle that?